lesson21/src/js/main.js

```diff
@@ -0,0 +1,212 @@
+        const fragment = document.createDocumentFragment();
+        const tr = document.createElement("tr");
+
+        Object.keys(usersTableColumn).forEach((column) => {
+            const td = createAttributedElements({
+                tag:"td",
+                valuesByAttributes:{
+                    class:"border border-gray-300 px-4 py-2"
+                },
+                str:user[column]
+            })
+            fragment.appendChild(tr).appendChild(td);
+        });
+
+        table.appendChild(tbody).appendChild(fragment);
+    }
+}
+
+const sortButtons = [
+    {dataOrder:"asc",src:"../img/asc.svg", alt:"asc-image"},
+    {dataOrder:"desc",src:"../img/desc.svg",alt:"desc-image"},
+    {dataOrder:"both",src:"../img/both.svg",alt:"both-image"}
+]
+
+const sortByUsersTableColumn = {
+    "id" : "ID"
+}
+
+const addSortButtons = (userKey,userValue,th) => {
+    for(const value of Object.values(sortByUsersTableColumn)){
+        userValue === value && createSortButtons(userKey,th) ;
+    }
+}
+
+const createSortButtons = (columnKey,th) => {
+    const sortButtonsBox = createAttributedElements ({
+        tag:"div",
+        valuesByAttributes:{
+            id:`js-${columnKey}Buttons-Box`,
+            class: "w-10 inline-block align-middle"
+        }
+    })
+
+    const fragment = document.createDocumentFragment();
+
+    for(const button of sortButtons){
+        const sortButton = createAttributedElements({
+            tag:"button",
+            valuesByAttributes:{
+                class: `js-${columnKey}SortButton`,
+                'data-order':button.dataOrder
+            }
+        });
+
+        button.dataOrder !== "both" && sortButton.classList.add("hidden");
+
+        const sortButtonImage = createAttributedElements({
+            tag:"img",
+            valuesByAttributes:{
+                class:"pointer-events-none",
+                src:button.src,
+                alt:button.alt
+            }
+        })
+
+        fragment.appendChild(sortButton).appendChild(sortButtonImage);
+    }
+    th.appendChild(sortButtonsBox).appendChild(fragment);
+}
+
+const toggleHiddenClassSortButton = (sortButtons,index) => {
+    sortButtons[index].classList.add("hidden");
+    (index === sortButtons.length -1) ? index = 0 : index++;
+    sortButtons[index].classList.remove("hidden");
+}
+
+const upDateTableBody = usersData =>  {
+    document.getElementById("js-usersTableBody").remove();
+    renderTableBody(usersData);
+};
+
+const sortUsersData = (key, usersData, dataOrder) => {
+    const cloneUsersData = [...usersData];
+
+    const sortButtonsFunc = {
+        "both": () => {
+            cloneUsersData.sort((a,b) => a[key] - b[key]);
+            upDateTableBody(cloneUsersData);
+        },
+        "asc" : () => {
+            cloneUsersData.sort((a,b) => b[key] - a[key]);
+            upDateTableBody(cloneUsersData);
+        },
+        "desc": () => {
+            upDateTableBody(usersData);
+        }
```

Comment:
IMO
I felt it would be more intuitive and easier to understand to have desc sort processing for `desc` and asc processing for `asc`.

Reply:
That is certainly true.
I fixed.
https://github.com/nakagawamai/js-lesson/pull/23/commits/a3c69dd7871347bc9a2092cc49707c95cfea29e3

Reply:
Wouldn't `sortButtonAttributes` be better left as is? If the sort order changes, for example, I think it would be difficult and costly to know where to modify.
What do you think about using a switch statement?
``` js
  switch (dataOrder) {
    case "asc":
      cloneUsersData.sort((a, b) => a[key] - b[key]);
      updateTableBody(cloneUsersData);
      break;
    case "desc":
      cloneUsersData.sort((a, b) => b[key] - a[key]);
      updateTableBody(cloneUsersData);
      break;
    case "both":
      updateTableBody(usersData);
      break;

    default:
    // default code
  }



```

Reply:
Thank you for your comment!
I think you are right, but if I leave it as it is,
Since it behaves differently from the specification, I rearranged it.

Clicked both button  → normal order
Clicked asc button → sort by id in ascending order
clicked desc button → sort by id in descending order
````JavaScript
const sortButtonAttributes = [
  { dataOrder: "desc", src: "../img/desc.svg", alt: "desc-image" },
  { dataOrder: "both", src: "../img/both.svg", alt: "both-image" },
  { dataOrder: "asc", src: "../img/asc.svg", alt: "asc-image" },
];
````

The instructions are
Click both button → Sort by id in ascending order
Click asc button → Sort by  id in descending order
Click desc button → normal order

So how should I solve this?

Reply:
Using the Switch statement is very smart!
Thank you for teaching.
I fixed.
https://github.com/nakagawamai/js-lesson/pull/23/commits/2c1229bdbe8e6c1f6ed5ca21fd2d7bb515a44818

Reply:
### [WIP]
indexの更新だけを行なってreturnし
``` js
const toggleHiddenClassSortButton = (sortButtons, index) => {
  // sortButtons[index].classList.add("hidden");
  index === sortButtons.length - 1 ? (index = 0) : index++;
  return index;
  // sortButtons[index].classList.remove("hidden");
};
```

``` js

const sortTableBody = (usersData) => {
  sortCategories.forEach((category) => {
    const sortButtons = [
      ...document.getElementsByClassName(`js-${category}SortButton`),
    ];
    const sortButtonsBox = document.getElementById(`js-${category}Buttons-Box`);

    // let dataOrder;
    sortButtonsBox.addEventListener("click", (e) => {
      if (e.currentTarget === e.target) return;

      const clickedSortButtonIndex = sortButtons.indexOf(e.target);
      // dataOrder = e.target.getAttribute("data-order");

      const nextOrderIndex = toggleHiddenClassSortButton(
        sortButtons,
        clickedSortButtonIndex
      );

      e.target.classList.add("hidden");
      sortButtons[nextOrderIndex].classList.remove("hidden");

      // toggleHiddenClassSortButton(sortButtons, clickedSortButtonIndex);
      sortUsersData(
        category,
        usersData,
        sortButtonAttributes[nextOrderIndex].state
      );
    });
  });
};

```

こんな感じにするとできると思います
※class名や関数の切り分け等は見直し必要だと思います


また、現状配列の順番に依存しているのが気になります。
``` js
const sortButtonAttributes = [
  { state: "both", src: "../img/both.svg", alt: "both-image" },
  { state: "asc", src: "../img/asc.svg", alt: "asc-image"},
  { state: "desc", src: "../img/desc.svg", alt: "desc-image"},
];

```

理由としては、この順番がソート順に影響を及ぼすかどうか
わかりづらいと感じる点、
`sortButtonAttributes`の名前とそわないと感じるからです。
(ちょっと私の感覚的な話だったらすいません、修正は任せます)

自分としてはswitch文で明示的に切り替える方がいいと思ってます

``` js
const hoge = (state) => {
  switch (state) {
    case "asc":
      return "desc";
    case "desc":
      return "both";
    default:
      return "asc";
  }
};

```

Reply:
詳しい解説、コードのご提案をいただき、ありがとうございます！
手詰まりになってしまい、大変助かりました。

たしかに、配列の順番に依存するのは、かなり問題ありますね。
sortButtonAttributesの内容を見ただけでは、順番を意識しないと思うので。

ボタンのhiddenクラスの切り替えをindexで処理している箇所が関係していそうなので、
依存しない方法に変えようと思います。

理解が遅くて大変申し訳ないですが、
もう一度よくオブジェクトの内容とボタンの切り替え部分を考え直してみます！

Reply:
ボタンの状態をswitch文で切り替える方法は、
大変わかりやすく良い方法だと思いましたので、修正致しました。
また、ボタンの表示を切り替える関数の処理も、ご提案頂いた配列の順番に依存しない方法に変更しました。

Thank you for all the advice.
Based on the comments we received, I have fixed them.

Renamed the property of sortButtonAttributes.
https://github.com/nakagawamai/js-lesson/pull/23/commits/a68446bb984fbbe45c03ea08063162a4c9b4f018

Added to switch to the next state function.
https://github.com/nakagawamai/js-lesson/pull/23/commits/d8564fc3d8dac5194e552619459a8679a7c4ae79
https://github.com/nakagawamai/js-lesson/pull/23/commits/a9c2ab245dccd2ecbe1ffdbb2093d2994d4dd0aa

Changed the processing for toggleHiddenClassForSortButton.
https://github.com/nakagawamai/js-lesson/pull/23/commits/de39d68fee98b91fb525b9e681ee49ed179a3136